drivers: sensor: tmp112: Add default device initialization

kurddt opened the pull request:
This allow to have a smaller driver footprint when a single
configuration is required.

Allow to have multiple driver instances

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

nashif commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -188,6 +194,41 @@ static int tmp112_channel_get(const struct device *dev,
 	return 0;
 }
 
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
```
The code here seems to be against an older version of Zephyr, the Kconfigs and some of the APIs have changed, please update against latest Zephyr tree.

kurddt replied:
I didn't saw that is config has been changed in favor to `CONFIG_PM_DEVICE`. I will also change it in my other open PR

gmarull commented on drivers/sensor/tmp112/Kconfig:
```diff
@@ -12,3 +12,42 @@ config TMP112
 	  The TMP102 is compatible with the TMP112 but is less accurate and has
 	  been successfully tested with this driver.
+
+choice
+	prompt "Conversion rate"
+	default TMP112_CR_RUNTIME
+	help
+	Initial conversion rate set at initialization
+	Supported values 0.25Hz, 1Hz, 4Hz, 8Hz
+
+config TMP112_CR_RUNTIME
+	bool "Set at runtime"
+
+config TMP112_CR_025HZ
+	bool "0.25Hz"
+
+config TMP112_CR_1HZ
+	bool "1Hz"
+
+config TMP112_CR_4HZ
+	bool "4Hz"
+
+config TMP112_CR_8HZ
+	bool "8Hz"
+endchoice
```
This setting should be part of devicetree, not Kconfig

kurddt replied:
True this would allow to have per device init values

avisconti replied:
Of course in DT to allow multi-instance

gmarull commented on drivers/sensor/tmp112/Kconfig:
```diff
@@ -12,3 +12,42 @@ config TMP112
+	bool "8Hz"
+endchoice
+
+choice TMP112_EXTENDED_MODE
+	prompt "Extended mode"
+	default TMP112_EM_RUNTIME
+	help
+	Initial extended mode (13-bits) set at initialization
+
+config TMP112_EM_RUNTIME
+	bool "Set at runtime"
+
+config TMP112_EM_ON
+	bool "On"
+
+config TMP112_EM_OFF
+	bool "Off"
```
Same here

gmarull commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -14,32 +14,28 @@
-
-/* scale in micro degrees Celsius */
-#define TMP112_TEMP_SCALE		62500
+static inline const struct tmp112_config *get_config(const struct device *dev)
+{
+	return dev->config;
+}
```
This function is redundant, `struct tmp112_config *config = dev->config` can be used instead

kurddt replied:
I used to have compilation warning when doing so let me check again

avisconti replied:
I'm not against the usage of inline functions like this, honestly. The code looks more clear and compiler should optimize anyway. But if you want to follow the suggestion I think the warning issue is maybe because the "const" dropped.

kurddt replied:
I also like those accessor function. they are present in some other device drivers as well

gmarull replied:
I think most of these helpers come from a period when Zephyr had _complicated names_ for the config/data fields in the device struct. I won't block on that, but I think new or refactored drivers should stop using such dummy functions.

gmarull commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -14,32 +14,28 @@
-	const struct device *i2c;
-	int16_t sample;
-};
+static inline struct tmp112_data *get_data(const struct device *dev)
+{
+	return dev->data;
+}
```
Same here

gmarull commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -188,6 +194,41 @@ static int tmp112_channel_get(const struct device *dev,
 
+#ifdef CONFIG_PM_DEVICE
+static int tmp112_pm_control(const struct device *dev, uint32_t command,
+			     uint32_t *context, pm_device_cb cb, void *arg)
```
```suggestion
			     uint32_t *state, pm_device_cb cb, void *arg)
```

gmarull commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -196,19 +237,29 @@ static const struct sensor_driver_api tmp112_driver_api = {
-		LOG_DBG("Failed to get pointer to %s device!",
-			    DT_INST_BUS_LABEL(0));
+	if (!device_is_ready(config->bus)) {
+		LOG_DBG("device %s not ready!", config->bus->name);
```
```suggestion
		LOG_ERR("device %s not ready!", config->bus->name);
```

gmarull reviewed (changes requested):
Thanks for these changes, added some comments.

avisconti reviewed (changes requested):
I see 3 different purposes in the same commit. Shouldn't it be splitted into at least 3 different commits with all suggestions received?

1. Add multi-instance support
2. Add fixed configuration at compile time
3. Add PM routines

What do you thi nk?

kurddt commented:
> I see 3 different purposes in the same commit. Shouldn't it be splitted into at least 3 different commits with all suggestions received? 
>     1. Add multi-instance support
>     2. Add fixed configuration at compile time 
>     3. Add PM routines 
> 
> What do you think?

I started doing that but since all three modify the newly introduced file `tmp112.h` it was not so straight forward to doing so. If you think It's a must have I can have a second attempt at it. It will probably be easier when I will switch from KConfig to DT

avisconti commented:
> > I see 3 different purposes in the same commit. Shouldn't it be splitted into at least 3 different commits with all suggestions received?
> > 1. Add multi-instance support
> > 2. Add fixed configuration at compile time
> > 3. Add PM routines
> > What do you think?
> 
> I started doing that but since all three modify the newly introduced file `tmp112.h` it was not so straight forward to doing so. If you think It's a must have I can have a second attempt at it. It will probably be easier when I will switch from KConfig to DT

Yes please. Let's give a try,

avisconti commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -98,8 +114,7 @@ static int tmp112_attr_set(const struct device *dev,
 
-		if (tmp112_reg_update(drv_data, TMP112_REG_CONFIG,
-				      TMP112_EM_BIT, value) < 0) {
+		if (tmp112_update_config(dev, TMP112_CONFIG_EM, value) < 0) {
```
Is this part (and related call) belonging to multi-instance commit? Or is it part of the configuration change?

kurddt replied:
It's actually a bit of both, I will try to reflect that in the commit separation

avisconti commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -133,9 +148,7 @@ static int tmp112_attr_set(const struct device *dev,
-		if (tmp112_reg_update(drv_data, TMP112_REG_CONFIG,
-				      TMP112_CR0_BIT | TMP112_CR1_BIT,
-				      value) < 0) {
+		if (tmp112_update_config(dev, TMP112_CONV_RATE_MASK, value) < 0) {
```
same this one

avisconti commented on drivers/sensor/tmp112/Kconfig:
```diff
@@ -12,3 +12,10 @@ config TMP112
+
+config TMP112_SAMPLING_FREQUENCY_RUNTIME
+	bool "Allow to set extended mode at runtime"
+
```
Putting this part here means that once selected all instance will be configurable at runtime. Is this what you want? I guess it is done for simplicity, and I'm not against it.

kurddt replied:
Yes that's true. This is done to safe some resources (flash) when the sensor configuration does not need to be changed after initialization which I think is the most common use-case. That said for backward compatibility it would be better to have it activated by default, like the older version. This has been lost in the rebase

avisconti commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -194,6 +194,41 @@ static int tmp112_channel_get(const struct device *dev,
+	return ret;
+}
+#endif
+
```
I was expecting tmp112_pm_control to be part of the driver definition.

kurddt replied:
done

avisconti reviewed (changes requested):
PLease take care also of putting the commit messages correctly (the multi-instance one is not correct)

avisconti reviewed (approved):
LGTM

gmarull commented on drivers/sensor/tmp112/Kconfig:
```diff
@@ -12,3 +12,11 @@ config TMP112
 	  The TMP102 is compatible with the TMP112 but is less accurate and has
 	  been successfully tested with this driver.
+
+config TMP112_FULL_SCALE_RUNTIME
+	bool "Allow conversion rate at runtime"
+	default y
+
+config TMP112_SAMPLING_FREQUENCY_RUNTIME
+	bool "Allow to set extended mode at runtime"
+	default y
```
nit: both are missing `help`

gmarull commented on dts/bindings/sensor/ti,tmp112.yaml:
```diff
@@ -8,3 +8,21 @@ description: |
+      - 4000
+      - 8000
+  full-scale:
+    description: Full scale in degree Celsus
```
nit: `Celsius`

gmarull commented on dts/bindings/sensor/ti,tmp112.yaml:
```diff
@@ -8,3 +8,21 @@ description: |
+
+properties:
+  sampling-rate:
+    description: Sampling rate in mHZ (milli-Hertz)
```
nit: `mHz`

gmarull commented on drivers/sensor/tmp112/tmp112.h:
```diff
@@ -44,6 +44,9 @@
 	int16_t sample;
 	uint16_t config_reg;
+#ifdef CONFIG_PM_DEVICE
+	uint32_t power_state;
```
this state is not initialized (should likely be set to active on init)

kurddt replied:
done

gmarull commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -194,6 +194,41 @@ static int tmp112_channel_get(const struct device *dev,
+			tmp112_update_config(dev, 0, 0);
+			device_data->power_state = PM_DEVICE_STATE_ACTIVE;
+		} else if (*state == PM_DEVICE_STATE_SUSPEND) {
+			device_data->power_state = PM_DEVICE_STATE_SUSPEND;
```
as far as I can see, only `PM_DEVICE_STATE_ACTIVE` is handled, if system requests `PM_DEVICE_STATE_SUSPEND` the driver does nothing but changing its state. How does device save power?

kurddt replied:
On it's own it does not. But if the device can be powered on/off by external electronics then the state can be restored.

gmarull replied:
A driver that implements PM needs to support at the very least active/suspend states. If the device can be powered on/off externally, then the device driver needs to know the actual device state, or, have control of such external electronics to change the device power states.

kurddt replied:
> then the device driver needs to know the actual device state

gmarull replied:
The problem I see here is that this is not actually implementing any device PM. If system is suspended, the device will get called with `PM_DEVICE_STATE_SET` and `PM_DEVICE_STATE_SUSPEND`. In this case, the device will do nothing other than setting it's state to `PM_DEVICE_STATE_SUSPEND`. As far as I understand, this is equivalent to have no device PM.

kurddt replied:
The driver send the last known configuration when the pm state is set from `PM_DEVICE_STATE_SUSPEND` to `PM_DEVICE_STATE_ACTIVE`. I record the current state in order to not send the configuration twice if the `PM_DEVICE_STATE_SET` is done twice in a row. We could save the configuration upon `PM_DEVICE_STATE_SUSPEND` but it's not required since `tmp112_attr_set` already does it

gmarull replied:
I'm sorry, but I still don't understand how the device enters a _low power state_ if the system is suspended. `tmp112_attr_set` has nothing to do with device PM. Can you please detail the workflow?

kurddt replied:
So let's say we have a board with several power domain. In one of those power domain, controlled by component `PWR`, we can find among other devices a `TMP112` device.

The `PWR` device driver (or the application) do know which components via (struct devices *) are under the power domain it controls. When the power domain is not in use anymore and need to be brought down, `TMP112` `pm_device_state_set(dev, PM_DEVICE_STATE_SUSPEND)` is called and then the power is turned-off.

When the power domain is required, i.e the application need access to one of the component. The power is turned on by `PWR` and  `pm_device_state_set(dev, PM_DEVICE_STATE_ACTIVE)` is called

gmarull replied:
I'd expect to see something related to this sensor feature in the PM section:

![image](https://user-images.githubusercontent.com/25011557/122373232-9bf16d00-cf61-11eb-946e-7bea406a6e54.png)

What you are describing is really an application specific behavior as far as I understand.

MaureenHelm commented on drivers/sensor/tmp112/Kconfig:
```diff
@@ -12,3 +12,17 @@ config TMP112
+	help
+	  When set extended mode can be selected using tmp112_attr_set
+	  with SENSOR_ATTR_FULL_SCALE and value to set either 128 or 150
+	default y
```
please move this line above the help section

MaureenHelm commented on drivers/sensor/tmp112/Kconfig:
```diff
@@ -12,3 +12,17 @@ config TMP112
+	help
+	  When set conversion rate can be set at runtime using sensor_attr_set
+	  with SENSOR_ATTR_SAMPLING_FREQUENCY
+	default y
```
please move this line above the help section

MaureenHelm commented on dts/bindings/sensor/ti,tmp112.yaml:
```diff
@@ -8,3 +8,21 @@ description: |
+  sampling-rate:
+    description: Sampling rate in mHz (milli-Hertz)
+    type: int
+    default: 1000
```
Where does this default come from? Is it the hardware reset value in the sensor?

kurddt replied:
I thought it was but after checking the hardware default is actually 4Hz. I will change that to align with hardware reset values

MaureenHelm commented on dts/bindings/sensor/ti,tmp112.yaml:
```diff
@@ -8,3 +8,21 @@ description: |
+  full-scale:
+    description: Full scale in degree Celsius
+    type: int
+    default: 128
```
ditto

kurddt replied:
Yes this is the hardware default

Commit:
drivers: sensor: tmp112: Add default device initialization

This allow to have a smaller driver footprint when a single
configuration is required.

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

MaureenHelm commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -48,38 +47,39 @@ static int tmp112_reg_read(struct tmp112_data *drv_data,
 {
-	struct tmp112_data *drv_data = dev->data;
-	int64_t value;
+	uint16_t value;
```
Is changing the type related to adding multi-instance support? If not, please make this change in a separate commit.

kurddt replied:
Moved to a separate commit

Commit:
drivers: sensor: tmp112: correct type

TMP112 registers are 16bit so there is no reason to use int64_t for
register values

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

MaureenHelm reviewed (commented):
Need to resolve PM questions raised by @gmarull , but otherwise LGTM.

kurddt commented:
@gmarull For some reason Github only allow me to resolve the conversation but not to add a reply. So I will answer here

You are right the shutdown feature could be used in the suspend state.

I don't think the PM implementation I propose is a application specific behavior. 
Without I don't think it's possible to have fine grain device power management with zephyr using device driver. 

That said I can remove the power management function and start a discussion on Slack about how to support such use case if you prefer.

gmarull commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -215,6 +214,8 @@ int tmp112_init(const struct device *dev)
 		.bus = DEVICE_DT_GET(DT_INST_BUS(n)),					 \
 		.i2c_slv_addr = DT_INST_REG_ADDR(n),					 \
+		.cr = DT_ENUM_IDX(DT_DRV_INST(n), sampling_rate),			 \
+		.extended_mode = DT_ENUM_IDX(DT_DRV_INST(n), full_scale),		 \
```
an enum property being used as a boolean sounds weird, why not use a boolean in devicetree?

kurddt replied:
That way the devicetree and `tmp112_attr_set(SENSOR_ATTR_FULL_SCALE)` are coherent, they accept the same values

gmarull commented on drivers/sensor/tmp112/tmp112.h:
```diff
@@ -8,14 +8,49 @@
+#define TMP112_CONV_RATE_4      2
+#define TMP112_CONV_RATE_8      3
+
+#define TMP112_CONV_RATE(cr)    (((uint16_t)(cr)) << TMP112_CONV_RATE_SHIFT)
```
cast should be performed outside of the macro

gmarull commented on drivers/sensor/tmp112/tmp112.c:
```diff
@@ -58,20 +46,27 @@ static int tmp112_reg_write(const struct device *dev,
+{
+	int rc;
+	struct tmp112_data *data = dev->data;
+	const uint16_t new_val = set_config_flags(dev, mask, val);
```
`const` redundant here

kurddt replied:
What do you mean ? The `const` is to indicate that `new_val` will not change after being set

gmarull commented on drivers/sensor/tmp112/Kconfig:
```diff
@@ -12,3 +12,17 @@ config TMP112
 	  The TMP102 is compatible with the TMP112 but is less accurate and has
 	  been successfully tested with this driver.
+
+config TMP112_FULL_SCALE_RUNTIME
+	bool "Allow to set extended mode at runtime"
+	default y
+	help
+	  When set extended mode can be selected using tmp112_attr_set
+	  with SENSOR_ATTR_FULL_SCALE and value to set either 128 or 150
+
+config TMP112_SAMPLING_FREQUENCY_RUNTIME
```
is it really needed to make these kconfig options, considering these are not per-instance?

kurddt replied:
This is in order to reduce the application footprint when those features are not required

gmarull replied:
ok, understood, thanks for the clarification

kurddt commented:
The PR will now probably conflicts now that #37750 has been recently merged

kurddt commented:
@MaureenHelm @gmarull Could you approve the PR if everything looks good to you ?

gmarull commented on dts/bindings/sensor/ti,tmp112.yaml:
```diff
@@ -8,3 +8,21 @@ description: |
+      - 1000
+      - 4000
+      - 8000
+  full-scale:
```
this property is being treated as a boolean (stored in `extended_mode`, which is `bool extended_mode: 1`). DT binding should reflect that as well.

gmarull commented on dts/bindings/sensor/ti,tmp112.yaml:
```diff
@@ -8,3 +8,21 @@ description: |
 include: i2c-device.yaml
+
+properties:
+  sampling-rate:
```
driver uses `cr` (conversion rate), can names be aligned?

kurddt commented:
@gmarull I addressed your latest requested changes

kurddt commented:
@nashif I would be nice if you can review again this PR in order to be merged
